### PR TITLE
Optimize dedup pipeline with metadata precheck and hash cache

### PR DESCRIPTION
**Pull request by Puneet-Maan**
## 📝 Description

Optimized deduplication performance in `src/storage/dedup_index.py` by adding a metadata precheck and hash cache so unchanged files are not re-hashed on every analysis run.

Previously, dedup always computed SHA-256 for every file in every run.  
Now, dedup stores per-file cache entries (`size`, `mtime_ns`, `hash`) in `dedup_index.json` under `__file_cache__` and only recomputes hash when metadata changes.

This keeps current dedup behavior and API outputs intact while reducing repeated I/O/CPU for incremental uploads.

**Closes:** #436 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [x] ⚡ Performance improvement

---

## 🧪 Testing

pytest test/test_dedup_index.py

Tests added in `test/test_dedup_index.py`:
- unchanged files reuse cached hash
- changed files are re-hashed
- deleted duplicate files are removed from __file_cache__

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [x] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

image

### CODE REVIEW

**Review by ThunderIW — CHANGES_REQUESTED** [review]
Caching logic looks good overall — clean implementation and backward compatible. Requesting changes for one item:              
                                                                                                                                 
Deleted file cache cleanup needs a test. The code removes cache entries when files are deleted, but there's no test covering   
  this. Without it, we can't be confident that the cache won't grow unbounded or return ] if unchecked I would recommed you add a test for this, once it is done, I will be happy to approve

File: test/test_dedup_index.py [hunk 1]
```diff
@@ -152,3 +186,57 @@ def test_lock_contention_times_out_and_warns(tmp_path, monkeypatch, caplog):
...
+    def counting_hash(path: Path) -> str:
+        calls["count"] += 1
+        return original(path)
+
+    monkeypatch.setattr(dedup_mod, "_file_hash", counting_hash)
+
+    deduplicate_project(proj, index_path)
+    assert calls["count"] == 1
+
+    f.write_text("v2")  # updates metadata, must trigger re-hash
```
[the remark this answers is not in the document]
Comment (ThunderIW) [reply test/test_dedup_index.py]: This a fix for the Docker test because it fails here  due to the fact the write happens so fast that the mtime (even at nanosecond resolution) doesn't change, so the cache thinks the
 The file is unchanged and skips re-hashing. but this purely a linux issue, feel free to change this if you want

**Comment by mahigangal** [comment]
You can reduce risk by including ctime_ns in the cache fingerprint (not just size + mtime_ns). Also, please update/add one test for this edge case.

**Comment by Puneet-Maan** [comment]
@mahigangal wouldn't the modified time change if the content is modified? and hence, it won't be a problem?

**Comment by Puneet-Maan** [comment]
@mahigangal regardless, it's a good practice! thank you for thinking of making the system more reliable. Implementing now
